config/config.go

```diff
@@ -651,31 +661,44 @@ func isTerragruntModuleDir(path string, terragruntOptions *options.TerragruntOpt
 // Read the Terragrunt config file from its default location
 func ReadTerragruntConfig(terragruntOptions *options.TerragruntOptions) (*TerragruntConfig, error) {
 	terragruntOptions.Logger.Debugf("Reading Terragrunt config file at %s", terragruntOptions.TerragruntConfigPath)
-	return ParseConfigFile(terragruntOptions.TerragruntConfigPath, terragruntOptions, nil, nil)
+
+	ctx := NewContext(context.Background(), terragruntOptions)
+	return ParseConfigFile(ctx, terragruntOptions.TerragruntConfigPath, nil)
 }
 
 // Parse the Terragrunt config file at the given path. If the include parameter is not nil, then treat this as a config
 // included in some other config file when resolving relative paths.
-func ParseConfigFile(filename string, terragruntOptions *options.TerragruntOptions, include *IncludeConfig, dependencyOutputs *cty.Value) (*TerragruntConfig, error) {
-	configString, err := util.ReadFileAsString(filename)
+func ParseConfigFile(ctx *Context, configPath string, includeFromChild *IncludeConfig) (*TerragruntConfig, error) {
+	// Parse the HCL file into an AST body that can be decoded multiple times later without having to re-parse
+	file, err := hclparse.NewParser().WithOptions(ctx.ParserOptions...).ParseFromFile(configPath)
 	if err != nil {
 		return nil, err
 	}
 
-	// Initialize evaluation context extensions from base blocks.
-	contextExtensions := &EvalContextExtensions{
-		DecodedDependencies: dependencyOutputs,
+	config, err := ParseConfig(ctx, file, includeFromChild)
+	if err != nil {
+		return nil, err
 	}
 
-	config, err := ParseConfigString(configString, terragruntOptions, include, filename, contextExtensions)
+	return config, nil
+}
+
+func ParseConfigString(ctx *Context, configPath string, configString string, includeFromChild *IncludeConfig) (*TerragruntConfig, error) {
```

Comment:
To put it another way: if you needed to add a totally _new / standalone_ package to parse _just_ the `catalog` block from a `terragrunt.hcl`, what would that look like? Perhaps that would take far fewer code changes and be more self-contained and easier to review?

Reply:
> Ooof, there are a lot of changes in `config.go`, `config_helpers.go`, etc... Very hard to review all of them. The logic in the `config` package is used for parsing all TG configs, and as these configs can be quite flexible, I'm a bit worried about missing some subtle issue amidst all these changes.

I completely share your concern. Luckily we have a lot of tests that are all related to configuration in one way or another. I'm almost sure that everything should work correctly, since all the tests passed.

> Is there any way to add this `catalog` parsing in a way that does _not_ affect parsing of everything else? I'd hate for the parsing of one new piece of functionality to potentially break the parsing of everything else...

> To put it another way: if you needed to add a totally new / standalone package to parse just the catalog block from a terragrunt.hcl, what would that look like? Perhaps that would take far fewer code changes and be more self-contained and easier to review?

I thought for a long time about how to avoid such changes, but unfortunately I don't see a better way out.

1. Writing a new separate parser purely for a `catalog` would be an even worse solution, since we would have to take into account all possible configurations where the `catalog` could be related with other nested configurations. Ultimately we will have to maintenance two version of reading configuration.
2. Since the code for reading the configuration is quite complex, and in certain situations it is somewhat recall of a recursive function and such variables as parsing options must be known throughout the entire process, and they cannot be static, implementing this task I tried not to complicate it even more by adding additional arguments to all functions.

Reply:
Understood.

In that case, @denis256, could I get a second set of eyes on this parsing code?